File: 알고리즘 문제풀이/kakao/실패율.py
from collections import defaultdict
import heapq
#실패율: 스테이지에 도달했으나 아직 클리어하지 못한 플레이어의 수 / 스테이지에 도달한 플레이어 수
#실패율이 같은 스테이지가 있다면 작은 번호의 스테이지가 먼저 오도록
#도달한 유저가 없는 경우 해당 스테이지의 실패율은 0
def solution(N, stages):
    answer,h,total,d = [],[],len(stages),defaultdict(int)
    for i in stages:
        d[i]+=1 #현재 stage 도달한 유저 집계
    for i in range(1,N+1):
        if total>0:
            heapq.heappush(h,[-(d[i]/total),i]) #실패율 계산
            total-=d[i] #이전 stage 도달한 유저수 만큼 차감
        else:
            heapq.heappush(h,[-0.0,i]) 
    while h:
        answer.extend([heapq.heappop(h)[1]])    
    return answer

File: 알고리즘 문제풀이/Algorithm/kakao/실패율.js
function solution(N, stages) {
  let answer = [];
  for (let i = 1; i <= N; i++) {
    let x = 0,
      y = 0;
    for (let j = 0; j < stages.length; j++) {
      if (stages[j] === i) x++;
      else if (stages[j] > i) y++;
    }
    if (x == 0 && y == 0) answer.push([i, 0]);
    else answer.push([i, x / (x + y)]);
  }
  answer = answer
    .sort((x, y) => (x[1] > y[1] ? -1 : x[1] < y[1] ? 1 : x[0] > y[0] ? 1 : -1))
    .map((item) => item[0]);
  return answer;
}
console.log(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]));
console.log(solution(4, [4, 4, 4, 4]));
